mh/session.py

```diff
@@ -215,3 +237,15 @@ def get_layer_players(self):
             return city.players
         else:
             assert False, "Can't find layer"
+            
+    def shutdown(self, shutdown_type):
+        assert 1 <= shutdown_type <= 2, "Invalid shutdown type"
+        self.graceful_shutdown = True
+        if shutdown_type != 1:
+            return  # only shutdown_type 1 requires removal from layers
+        if self.local_info["city_id"]:
```

Comment:
This logic is shared with `session.layer_end`, it's probably better to move it to its own function like `clear_local_info`.